targettests/execution/quantum_struct_exec.cpp

```diff
@@ -0,0 +1,37 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+// RUN: nvq++ %cpp_std --enable-mlir %s -o %t && %t | FileCheck %s
+
+#include "cudaq.h"
+
+struct test {
+  int i;
+  double d;
+  cudaq::qview<> q;
+};
+
+__qpu__ void hello(cudaq::qubit &q) { h(q); }
+
+__qpu__ void kernel(test t) {
+  h(t.q);
+  for (int i = 0; i < t.i; i++)
+    hello(t.q[i]);
+}
+
+__qpu__ void entry(int i) {
+  cudaq::qvector q(i);
+  test tt{i, 2.2, q};
+  kernel(tt);
+}
```

Comment:
Thanks, Alex. I downloaded this example to play with it a bit more. I can't get it to work for any of the target hardware. In essence, it looks like the loop unrolling passes can't figure out what to do with the `cudaq::qview<> q` struct member (or perhaps other members too).

Trying to simplify as much as possible ... if you remove the device `kernel` function and the `for` loop from the example, you're left with something like this:
```cpp
__qpu__ void entry(int i) {
  cudaq::qvector q(i);
  test tt{i, 2.2, q};

  // This works
  h(q);
  
  // This fails with error: 'cc.loop' op not a constant upper bound loop
  h(tt.q);
}
```
It seems like the above example should, in principle, be able to work, but when compiling and running with `nvq++ --target quantinuum --emulate ../targettests/execution/quantum_struct_exec.cpp`, it fails with the loop unrolling error.

Do you think there is a reasonable path forward to making this work? Would the constant propagation logic need to be updated to handle this? Or something else?

Reply:
I've added a canonicalizer to address this. Loops unroll now.

Reply:
Just to document this: few of the transformation passes were built to work with aggregate typed values. (None to work with loading and storing quantum typed reference values.)